.github/scripts/trymerge.py

```diff
@@ -986,38 +998,18 @@ def find_matching_merge_rule(pr: GitHubPR,
 def get_land_checkrun_conclusions(org: str, project: str, commit: str) -> Dict[str, Tuple[str, str]]:
     land_check_info = gh_get_land_check_info(org, project, commit)
     checksuites = land_check_info["checkSuites"]
-    conclusions = {}
-
-    def add_conclusions(edges: List[Dict[str, Dict[str, Any]]]) -> None:
-        for edge_idx, edge in enumerate(edges):
-            node = edge["node"]
-            workflow_run = node["workflowRun"]
-            checkruns = node["checkRuns"]
-            if workflow_run is not None:
-                conclusions[workflow_run["workflow"]["name"]] = (node["conclusion"], node["url"])
-            has_failing_check = False
-            while checkruns is not None:
-                for checkrun_node in checkruns["nodes"]:
-                    if checkrun_node["conclusion"] == 'FAILURE':
-                        has_failing_check = True
-                    conclusions[f'{get_check_run_name_prefix(workflow_run)}{checkrun_node["name"]}'] = (
-                        checkrun_node["conclusion"], checkrun_node["detailsUrl"]
-                    )
-                if bool(checkruns["pageInfo"]["hasNextPage"]):
-                    rc = gh_graphql(GH_GET_COMMIT_NEXT_CHECK_RUNS,
-                                    name=project,
-                                    owner=org,
-                                    cs_cursor=edges[edge_idx - 1]["cursor"] if edge_idx > 0 else None,
-                                    cr_cursor=checkruns["pageInfo"]["endCursor"],
-                                    commit=commit)
-                    checkruns = rc["data"]["repository"]["object"]["checkSuites"]["nodes"][-1]["checkRuns"]
-                else:
-                    checkruns = None
-            # Github doesn't set conclusion to failure if a job is still pending
-            if workflow_run is not None and has_failing_check:
-                conclusions[workflow_run["workflow"]["name"]] = ("FAILURE", node["url"])
+    conclusions: Dict[str, Tuple[str, str]] = {}
+
+    def get_commit_next_check_runs(edges: List[Dict[str, Dict[str, Any]]], edge_idx: int, checkruns: Any) -> Any:
+        rc = gh_graphql(GH_GET_COMMIT_NEXT_CHECK_RUNS,
+                        name=project,
+                        owner=org,
+                        cs_cursor=edges[edge_idx - 1]["cursor"] if edge_idx > 0 else None,
+                        cr_cursor=checkruns["pageInfo"]["endCursor"],
+                        commit=commit)
+        return rc["data"]["repository"]["object"]["checkSuites"]["nodes"][-1]["checkRuns"]
 
-    add_conclusions(checksuites["edges"])
+    add_workflow_conclusions(conclusions, checksuites["edges"], get_commit_next_check_runs)
     while bool(checksuites["pageInfo"]["hasNextPage"]):
         rc = gh_graphql(GH_GET_COMMIT_NEXT_CHECKSUITES,
```

Comment:
can this pagination method be extracted out as well?

Reply:
Do you mean extract into a completely separate function or extract it so that its inside of add_workflow_conclusions?  If is the second one, get_commit_next_check_runs and its counterpart get_pr_next_check_runs are slightly different which is why add_workflow_conclusions takes in a callable

Reply:
Like the get_checkrun_conclusions/get_land_checkrun_conclusions still look kinda similar. Like we can maybe create another function that takes in the 2 callables to fetch the next page for checksuites/checkruns and have these call this new function.

Reply:
done